From 723b51878dfb7647e2ee11411bb8ac3cf49f584f Mon Sep 17 00:00:00 2001
From: vladimir.a.ivanov <vladimir.a.ivanov@intel.com>
Date: Fri, 29 Mar 2013 09:47:56 +0700
Subject: Dalvik: skip code registration if placeholder was cleared in the JitTable

BZ: 93808

In the case when code cache buffer reset occurred between events "place is
locked in the JitTable" and "trace compilation is finished" no JitEntry will
be available in the table. In this case no code register in the table.

Category: device-enablement
Domain: AOSP-Dalvik-Interpreter
Origin: internal
Upstream-Candidate: yes

Change-Id: I624e4099618da645dc702f0ff0fc4298f198fcc6
Orig-MCG-Change-Id: If272ab1ed6cf9924455a3dc22e0359f5febefbe3
Signed-off-by: vladimir.a.ivanov <vladimir.a.ivanov@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/interp/Jit.cpp |   15 ++++++++++++++-
 1 files changed, 14 insertions(+), 1 deletions(-)

diff --git a/vm/interp/Jit.cpp b/vm/interp/Jit.cpp
index 8ed474c..b225262 100644
--- a/vm/interp/Jit.cpp
+++ b/vm/interp/Jit.cpp
@@ -1136,7 +1136,20 @@ void dvmJitSetCodeAddr(const u2* dPC, void *nPC, JitInstructionSetType set,
     JitEntry *jitEntry = isMethodEntry ?
         lookupAndAdd(dPC, false /* caller holds tableLock */, isMethodEntry) :
                      dvmJitFindEntry(dPC, isMethodEntry);
-    assert(jitEntry);
+    /*
+     * The 'NULL' is valid return code for both methods 'lookupAndAdd' and
+     * 'dvmJitFindEntry'. But 'lookupAndAdd' insert missed address if no one in
+     * the table and return 'NULL' value for full table only while method
+     * 'dvmJitFindEntry' does not implement 'add' functionality.
+     * In the case when code cache buffer reset occurred between events "place is
+     * locked in the JitTable" and "trace compilation is finished" no JitEntry will
+     * be available in the table. In this case method dvmJitFindEntry returns NULL
+     * and no code is registering.
+     */
+    if (jitEntry == NULL) {
+        return;
+    }
+
     /* Note: order of update is important */
     do {
         oldValue = jitEntry->u;
-- 
1.7.4.1

